File: src/fishing.js
import { ButtonStyleTypes } from 'discord-interactions';

export function generateFish() {
  var fishList = ['🐟', '🐡', '🐠', '🦈', '🐳', '🐬', '🦀', '🦐', '🦑'];

  var fish = [];
  for (var i = 0; i < 5; i++) {
    var randomFish = fishList[Math.floor(Math.random() * fishList.length)];
    while (fish.includes(randomFish)) {
      randomFish = fishList[Math.floor(Math.random() * fishList.length)];
    }
    fish.push(randomFish);
  }

  return fish;
}

export function generateFishButtons(fishList, userId) {
  var fishActions = [];

  for (var i = 0; i < fishList.length; i++) {
    fishActions.push({
      type: 2,
      label: '',
      style: ButtonStyleTypes.PRIMARY,
      custom_id: userId + '-' + i,
      emoji: {
        name: fishList[i],
      },
    });
  }

  return fishActions;
}

File: fishing.go
package main

import (
	"fmt"
	"math/rand"
	"time"

	"github.com/bwmarrin/discordgo"
)

func generateFish() []string {
	fishList := []string{"🐟", "🐡", "🐠", "🦈", "🐳", "🐬", "🦀", "🦐", "🦑"}
	var fish []string

	for i := 0; i < 5; i++ {
		randomFish := fishList[rand.Intn(len(fishList))]
		for contains(fish, randomFish) {
			randomFish = fishList[rand.Intn(len(fishList))]
		}
		fish = append(fish, randomFish)
	}

	return fish
}

func contains(slice []string, val string) bool {
	for _, item := range slice {
		if item == val {
			return true
		}
	}
	return false
}

func generateFishButtons(fish []string, userId string) []discordgo.MessageComponent {
	var buttons []discordgo.MessageComponent
	var actionRowComponents []discordgo.MessageComponent

	for i, v := range fish {
		button := discordgo.Button{
			Emoji: discordgo.ComponentEmoji{
				Name: v,
			},
			CustomID: fmt.Sprintf("fish-%s-%d", userId, i),
			Style:    discordgo.PrimaryButton,
		}
		actionRowComponents = append(actionRowComponents, button)
	}

	buttons = append(buttons, discordgo.ActionsRow{
		Components: actionRowComponents,
	})

	return buttons
}

func fishButtonHandler(s *discordgo.Session, i *discordgo.InteractionCreate, message *discordgo.Message, sleep int, fish string) {
	sleeptime := time.Duration(sleep) * time.Second
	time.Sleep(sleeptime)

	messageText := fmt.Sprintf("React with %s!", fish)

	_, err := s.FollowupMessageEdit(i.Interaction, message.ID, &discordgo.WebhookEdit{
		Content:         &messageText,
		Components:      nil,
		AllowedMentions: &discordgo.MessageAllowedMentions{},
	})
	if err != nil {
		fmt.Println(err)
	}
}
